README
Sman
==================

NAME
	The sman README

What is sman?
	Sman is the Searcher for Man pages. Based on the example of the
	same name in Josh Rabinowitz's article "How To Index Anything"
	in the July, 2003 issue of Linux Journal 
	(http://www.linuxjournal.com/article.php?sid=6652), sman is
	an enhanced version of 'apropos' and 'man -k'.  Sman adds 
	several key abilities over its predecessors:

	* Supports complex natural language text searches such as 
		"(linux and kernel) or (mach and microkernel)"

	* Shows results in a ranked order, and optionally an extract 
       (using -e) of the manpage showing the searched text highlighted

	* Allows for searches by manpage section, title,
		body, or filename (use 'metaname=searchword')

	* Indexes the complete contents of the man page, not just 
		the title and description

	* Uses a prebuilt index to perform fast searches

	* Performs 'stemming' so that a search for "searches" 
	  will match a document with the word "searching"

Where do I get sman?  
	Sman is currently distributed from http://joshr.com/src/sman.

	It's always best to make sure you use a current version of sman.

How do I install sman?

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

Dependencies:
	Sman depends on SWISH-E 2.4 or above.

	SWISH-E can be found at http://www.swish-e.org/Download/
	
Building the index:
	The included sman-update script is used to build (and rebuild)
	the sman index, which is comprised of the files 'sman.index' 
	and 'sman.index.prop'.

Searching the man page index with sman:
	The sman command is fairly simple. As its help output says:

    Usage: sman [--max=#] [--rank] [--number] [--index='index'] 
                [--file] [--help] [--repeats] [--begin=#'] [--config=file]
                [--quote='"'] [--verbose] [--VERSION] searchword [...]
    Ranked freetext searches on manpages.
    Options:
      --max=#:                limit number of results, default 20
      --rank:                 show the rank of each hit
      --number:               show the number of each hit
      --index=index:          specify an index (overrides config)
      --file:                 show the source man file for each hit
      --help:                 this help information
      --repeats:              show repeat manpages
      --begin=#:              start showing hits at number N
      --config=my-sman.conf:  a config file (specs an index file)
      --quote='"':            specify a quoting char for output
      --extract:              show extraction of manpage for each hit
      --verbose:              show more output
      --VERSION:              show version and exit

MORE INFO
    Sman is thoroughly documented. For more info, see 'perldoc sman',  
    'perldoc sman-update', or the Sman distribution itself.

TESTING 
	Sman has been tested on several versions of perl and unix-like operating
    systems.  We welcome bug reports and fixes for all platforms!

COPYRIGHT AND LICENCE 
	This software is licensed the same as perl itself.  
	Copyright (C) 2003-2006 Josh Rabinowitz <joshr>



Changes
Revision history for Sman.  
http://joshr.com/src/sman/
Copyright 2004-2008 Josh Rabinowitz  <joshr>
#$Id: Changes,v 1.88 2008/06/03 15:26:46 joshr Exp $

1.03_001
        Changes for PAUSE, and changes for swish-e 2.6.
        
1.03   Wed May 28 2008
        Look in /usr/local/lib/swish-e/perl and /usr/libexec/swish-e/perl/
            for SWISH::DefaultHighlight. Latter is where rpms put it.
        Better error outputs if we can't locate swish-e or load SWISH::API.
        No known machines use 'cat %F | gunzip -f --stdout | man -c' 
          to get ascii from manpages; method removed (it caused warnings on CentOS4).
        Added to and changed the TODO file.
        Enable (undocumented) ability to use doclifter as a manpage-to-xml converter.
        Added --scheme=[0|1] command line option to sman (default is 1, if the
          version of Swish-e installed supports it).
        Add support for fink on MacOSX.
        Support for debian 4.0 and later.
        Code cleanups and corrections; better error reporting.
1.02 
        Not released
1.01   Thu Jul 06 2006
        Documentation updates
        Tiny change to sman.conf.pm's pod for Test::Pod
		Changes to sman-defaults.conf for swished's swished_example.cgi
		  (see http://search.cpan.org/dist/swished/ )
        Updates to TODO file
1.00   Thu May 04 2006
        Removed spurious error message when using sman's -v option without a query
        Changed source & -V version string to show Swish-e capitalized as here
        Test current Swish-e version more carefully in Sman::Util
        More error output in sman's undocumented --Debug mode
        Cleanup how we check for $SWISH::API::VERSION
        Touchups to formatting of Changes file
        Updated various dates to 2006
        Tiny POD touchups
0.99   Tue May 02 2006
        Avoid warnings from sman-update for nonexistant dirs in manpath
        Be more careful about configuration files and temp files
        Changes for win32 under cygwin
        Typo fixes and rewordings in docs, more TODO changes
        Little changes to sman-defaults.conf
        Added dependencies to Makefile.PL
        POD error in sman.conf.pm fixed as per cpants
        Tests added to test POD and POD coverage
0.98   Tue Aug 30 2005
        bugfix: added Sman/IndexVersion.pm to MANIFEST (thanks to testers!)
        added t/34_SmanIndexVersion.t to test Sman::IndexVersion
        TODO modified
0.97   Mon Aug 29 2005
        Check that a sufficient SWISH::API version is installed,
          handle absence better
        Fixed Sman::IndexVersion, it now uses Sman::Config
        Handle errors in sman better
        Added --progress and --max options to sman-update
        Documentation updates
0.96   Fri Aug 26 2005
        Choose number of results to show based on rows returned from Term::Size
          (to override use sman's --max=N command line option)
        Put quotes around verbatim error string from Swish-e if we show one
        Avoid warnings from zcat during sman-update autoconfiguration on OSX 10.4
        Use 'man -c' (and ENV_MANWIDTH) to actually wrap manpages to 256 char lines
        Fix NAME section of sman-update perldocs (to fix problem of 
          sman-update not appearing at http://search.cpan.org/~joshr/Sman-0.95/ )
        Document running as low-priv user in sman-update
        Update TODO and sman.conf documentation
        Show index update time in sman and sman-update --VERSION output
        Added /var/lib/sman/sman.index.version file and usage
        Use Swish-e's rank scheme 1 by default if index allows
0.95   Sat Jun 25 2005
        Changes for Swish-e deprecation of 'FuzzyIndexingMode Stem'
        Changed default to use 'zcat -f --stdout', not 'zcat -f'
        Changed -V output slightly
        More changes to tests
        Doc and typo fixes
        Avoid various warnings from different perls/OSs/helper programs
0.94   Wed Jan 05 2005 
        Renumbered versions from 0.9.x to 0.9x
        Fixed perldoc for Sman.pm 
        Prefix warnings from sman and sman-update with 'sman[-update]: '
        Have sman mention sman-update if the index file appears empty
0.9.3  Tue Jan 04 2005
        Moved perldocs from script/sman to Sman.pm
0.9.2  Tue Jan 04 2005
        Updated README to mention sman's --extract option
        Updated Usage in README and sman
        Change example in sman docs to use --rank instead of -r
        Show total hits and less about config files in sman -v
        Added Swish-e WordCharacters, BeginCharacters, and EndCharacters
         settings to sman-default.conf so underscores and colons are
         accepted as word characters, thus (for example) mod_perl and 
         Apache::Test are each indexed as single words
        use length(content) & 'use bytes' instead of pack() 
        Changed tests to succeed even if there is no /usr/local/lib/swish-e/ directory
0.9.1  Sun Jun 06 2004
        Updated README to reflect current output of 'sman -h'
        First Sman release on CPAN
0.9.0  Sun Jun 06 2004
        sman: added --extract option, which shows extract of manpage,
         use ResultPropertyString() to avoid exceptions, 
         make default width 80 chars if not connected to a terminal
        sman-update: only print errors from 'man' if -warn is enabled
        Added 'manpage' and 'digest' metas and props to index
        Added basic tests for sman and sman-update scripts
        Updated TODO file
        This version was never released
0.8.9  Mon May 31 2004
        Updated TODO, fixed copyright year and typos in README & sman-defaults.conf
        Get version string for perl, not its API version 
          in Sman::Util::GetVersionString
        Consider output of choice of MANCMD verbose 
0.8.8  Mon Apr 19 2004
        Disable warnings in some of Sman::Man::Cache::FileCache for perl 5.8.0
        Added perl version to -V(ERSION) output, show versions as '5.8.3'
        Updated README

Sman/sman.conf.pm
#!/usr/bin/perl -w
use strict;

#$Id: sman.conf.pm,v 1.7 2008/06/03 15:38:48 joshr Exp $

# this file exists solely so that you can type
# perldoc 'sman.conf' or 'man sman.conf' and get 
# meaningful data.  

=head1 NAME

sman.conf - configuration file for sman and sman-update

=head1 SYNOPSIS

The sman.conf configuration file specifies several run-time options
for sman and sman-update.

=head1 DESCRIPTION 

sman.conf is the default name of the configuration file 
used for sman.  Sman looks for a file called sman.conf in several 
usual places, and if not found, falls back on sman-defaults.conf, 
likely from the default location of /usr/local/etc.  See 
L<Sman::Config> for details.

=head1 SMAN CONFIG FILE

Each line in a sman.conf file is either a comment or a directive.  
Comments are lines that begin with a # character.  
Directives are of the form:
    
    Directive values values values ...

The directives currently understood in an sman configuration file are:

=head2 COLCMD

A program that is used to strip out backspaces and 
such from the MANCMD output. If undefined, defaults
to 'col -b'

    Example: COLCMD col -b

=head2 MANCMD

Program that first converts the troff to ASCII. 
Absence of a MANCMD directive, or the setting AUTOCONFIG, causes
sman-update to try to ascertain the best man command on its own.
MANCMD understands a few format sequences that are used to specify
how to get man to convert your manfiles on your system.  The format sequences 
currently understood are:
  %F   the full pathname of the man file (ie, /usr/share/man/man1/ls.1.gz)
  %C   the command name (ie, ls)
  %S   the section (ie, 1 or 3pm)

  Example: MANCMD AUTOCONFIG (will choose one of the below)
or
  Example: MANCMD man %F     (linux likes this)
or
  Example: MANCMD man %S %C  (freebsd prefers this)

Also note that if you have a custom mechanism to translate
man files into XML, you can use it like this: 

  Example: MANCMD zcat --stdout -f %F | myprog

=head2 SWISHECMD
    
The path and options you'd like to use with Swish-e while indexing. 
If undefined, will default to 'swish-e'. Appropriate Swish-e options 
will be appended when indexing.

    Example: SWISHECMD swish-e -v 1

=head2 TMPDIR

where to put various temporary files. Defaults to /tmp.
(Use SWISHE_TMPDIR to set affect Swish-e at index time) 

    Example: TMPDIR /tmp 

=head2 TITLEALIASES, SECALIASES, DESCALIASES

Aliases for the XML tags we expect. These are in case an external program
returns tags different from what we expect.

    Examples:
    TITLEALIASES refentrytitle
    SECALIASES manvolnum
    DESCALIASES refpurpose 
    MANPAGEALIASES swishdefault

=head2 ENV_* (ENV_MANWIDTH, ...)

All parameters beginning with ENV_ have the ENV_ prefix stripped 
and are used to set environment variables of the corresponding names
for child processes.

In particular, on some versions of man, the MANWIDTH environment variable
controls the desired line width for manpage output. This can affect 
line-breaking during indexing on some systems.

    Example:
    ENV_MANWIDTH 256

=head2 SWISHE_* (SWISHE_IndexFile, ...)

All parameters beginning with SWISHE_ have the SWISHE_ prefix stripped 
and are written into a tmp config file for Swish-e at index time.

SWISHE_IndexFile is also used by sman to know which index to search.

You shouldn't need to change any of the other SWISHE_* parameters.
They are all documented (without the SWISHE_ prefix) in the Swish-e
documentation.

    Examples: 

    # SWISHE_IndexFile specifies which index to create and search
    SWISHE_IndexFile /var/lib/sman/sman.index 
    SWISHE_IndexComments      no 
    # SWISHE_UseStemming       yes   # for old versions of Swish-e
    SWISHE_FuzzyIndexingMode   Stem
    SWISHE_MetaNames          desc sec swishtitle 
    SWISHE_PropertyNames      desc sec 

=head1 AUTHOR

Josh Rabinowitz <joshr>

=head1 SEE ALSO

L<sman>, L<sman-update>

1;
